File: MangaTracker/src/com/camlinard/mangatracker/BookLookup.java
package com.camlinard.mangatracker;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.MalformedURLException;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.StatusLine;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;
import org.json.JSONArray;
import org.json.JSONObject;

import android.os.AsyncTask;

public class BookLookup {
	static public final String kKey = "&key=AIzaSyDeQMfnPyhQ23-ndhb9xs9IY_EaSiTxgms";
	static public final String kURL = "https://www.googleapis.com/books/v1";
	static public final String kVolumesCollection = "volumes";
	static public final String kISBNParameter = "isbn:%s";
	static public final String kKind = "kind";
	static public final String kBooksVolumes = "books#volumes";
	static public final String kISBN = "isbn";
	static public final String kItemCount = "totalItems";
	static public final String kItems = "items";
	
	interface LookupDelegate {
		public void BookLookupResult(Book[] result, boolean more);
		public void BookLookupError(String error);
	}
	
	LookupTask m_lookup = null;
	
	public boolean Lookup(LookupDelegate results, String collection, String parameters) {
		if (m_lookup != null)
			return false;
		m_lookup = new LookupTask(results);
		m_lookup.execute(String.format("%s/%s?q=%s%s", kURL, collection, parameters, kKey));
		return true;
	}
	
	public boolean LookupISBN(LookupDelegate results, String isbn) {
		return Lookup(results, kVolumesCollection, String.format(kISBNParameter, isbn));
	}
	
	protected class LookupTask extends AsyncTask<String, Void, Book[]> {
		protected String m_url = null;
		protected LookupDelegate m_results;
		
		public LookupTask(LookupDelegate results) {
			super();
			m_results = results;
		}

		@Override
		protected Book[] doInBackground(String... params) {
			String spec = params[0];
			HttpClient bookClient = new DefaultHttpClient();
			try {
				HttpGet get = new HttpGet(spec);
				HttpResponse response = bookClient.execute(get);
				StatusLine status = response.getStatusLine();
				if (status.getStatusCode() != 200)
					throw new Exception(String.format("HTTP Error %d", status.getStatusCode()));

				HttpEntity entity = response.getEntity();
				InputStream content = entity.getContent();
				InputStreamReader input = new InputStreamReader(content);
				BufferedReader reader = new BufferedReader(input);
				StringBuilder responseBuilder = new StringBuilder();
				String lineIn;
				while ((lineIn=reader.readLine())!=null) {
					responseBuilder.append(lineIn);
				}

				String responseString = responseBuilder.toString();
				JSONObject json = new JSONObject(responseString);
				Book[] books = parseResponse(json);
				return books;
			} catch (MalformedURLException e) {
				m_results.BookLookupError(String.format("Error: %s: %s", e.toString(), params[0]));
			} catch (Exception e) {
				m_results.BookLookupError(String.format("Error: %s", e.toString()));
			}
			return null;
		}
		
		@Override
		protected void onPostExecute(Book[] result) {
			super.onPostExecute(result);
			m_results.BookLookupResult(result, false);
			m_lookup = null;
		}

		Book[] parseResponse(JSONObject json) throws Exception {
			String kind = json.getString(kKind);
			if (kind.equals(kBooksVolumes)) {
				int count = json.getInt(kItemCount);
				if (count == 0)
					return null;
				JSONArray items = json.getJSONArray(kItems);
				Book[] books = new Book[count];
				for (int i = 0; i < count ; ++i) {
					books[i] = Book.parseJSON(items.getJSONObject(i));
				}
				return books;
			}
			throw new Exception("Invalid Response");
		}
	}
}

File: BiblioTech/app/src/main/java/com/example/cleve/bibliotech/BookLookup.kt
package com.example.cleve.bibliotech

import android.os.AsyncTask
import org.json.JSONObject
import java.io.BufferedInputStream
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.lang.ref.WeakReference
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

internal class BookLookup {
    internal interface LookupDelegate {
        fun bookLookupResult(result: Array<Book?>?, more: Boolean)
        fun bookLookupError(error: String?)
    }

    private var mLookup: LookupTask? = null
    private fun lookup(
        results: LookupDelegate,
        collection: String,
        parameters: String
    ): Boolean {
        if (mLookup != null) return false
        mLookup = LookupTask(this, results)
        mLookup!!.execute(
            String.format(
                "%s/%s?q=%s%s",
                kURL,
                collection,
                parameters,
                kKey
            )
        )
        return true
    }

    fun lookupISBN(results: LookupDelegate, isbn: String?): Boolean {
        return lookup(
            results,
            kVolumesCollection,
            String.format(kISBNParameter, isbn)
        )
    }

    private class LookupTask internal constructor(parent: BookLookup, val mResults: LookupDelegate) :
        AsyncTask<String?, Void?, Array<Book?>?>() {
        val mParent = WeakReference(parent)
        override fun doInBackground(vararg params: String?): Array<Book?>? {
            val spec = params[0]
            var bookClient: HttpURLConnection? = null
            try {
                val url = URL(spec)
                bookClient = url.openConnection() as HttpURLConnection
                val content: InputStream =
                    BufferedInputStream(bookClient.inputStream)
                val status = bookClient.responseCode
                if (status != 200) throw Exception(
                    String.format(
                        "HTTP Error %d",
                        status
                    )
                )
                val input = InputStreamReader(content)
                val reader = BufferedReader(input)
                val responseBuilder = StringBuilder()
                var lineIn: String?
                while (reader.readLine().also { lineIn = it } != null) {
                    responseBuilder.append(lineIn)
                }
                val responseString = responseBuilder.toString()
                val json = JSONObject(responseString)
                return parseResponse(json)
            } catch (e: MalformedURLException) {
                mResults.bookLookupError(
                    String.format(
                        "Error: %s: %s",
                        e.toString(),
                        params[0]
                    )
                )
            } catch (e: Exception) {
                mResults.bookLookupError(String.format("Error: %s", e.toString()))
            } finally {
                bookClient?.disconnect()
            }
            return null
        }

        override fun onPostExecute(result: Array<Book?>?) {
            super.onPostExecute(result)
            mResults.bookLookupResult(result, false)
            val parent = mParent.get()
            if (parent != null)
                parent.mLookup = null
        }

        @Throws(Exception::class)
        fun parseResponse(json: JSONObject): Array<Book?>? {
            val kind = json.getString(kKind)
            if (kind == kBooksVolumes) {
                val count = json.getInt(kItemCount)
                if (count == 0) return null
                val items = json.getJSONArray(kItems)
                val books: Array<Book?> = arrayOfNulls(count)
                for (i in 0 until count) {
                    books[i] = Book.parseJSON(items.getJSONObject(i))
                }
                return books
            }
            throw Exception("Invalid Response")
        }

    }

    companion object {
        private const val kKey = "&key=AIzaSyDeQMfnPyhQ23-ndhb9xs9IY_EaSiTxgms"
        private const val kURL = "https://www.googleapis.com/books/v1"
        private const val kVolumesCollection = "volumes"
        private const val kISBNParameter = "isbn:%s"
        private const val kKind = "kind"
        private const val kBooksVolumes = "books#volumes"
        private const val kItemCount = "totalItems"
        private const val kItems = "items"
    }
}